### PR TITLE
Adds CMK for CosmosDB Account

### LINKED ISSUE #7798: Support for CosmosDB encryption with CMK
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

CosmosDB accounts now support encryption using customer-managed keys.
This is enabled as shown in the documentation here: https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-setup-cmk#create-a-new-azure-cosmos-account

It should be possible to do this with Terraform also.

### New or Affected Resource(s)

* azurerm_cosmosdb_account

### Potential Terraform Configuration

Probably this should be provided in a block like:

```hcl
resource "azurerm_cosmosdb_account" "db" {
  ...
  encryption {
    key_uri = "https://<my-vault>.vault.azure.net/keys/<my-key>"
  }
  ...
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by petems**
* Closes #7798

### CODE REVIEW

**Review by jackofallops — DISMISSED** [review]
Hi @petems - Thanks for this PR, looks off to a good start. 

My understanding is that the `key_vault_key_id` property can be changed / updated, so I think we'd need to remove the `ForceNew` and include it in the update. As it stands, I believe it will remove the key from the account was otherwise updated? Could you add a test that covers the new property also?

Thanks!

**Comment by katbyte** [comment]
Adding to @jackofallops's comment, could we also add a test that updates the `key_vault_key_id`

**Comment by petems** [comment]
> My understanding is that the `key_vault_key_id` property can be changed / updated, so I think we'd need to remove the `ForceNew` and include it in the update. As it stands, I believe it will remove the key from the account was otherwise updated? Could you add a test that covers the new property also?

Currently the `key_vault_key_id` cannot be changed in the API, its only possible on new resources, which is why I set it to `ForceNew`: 

`Code="BadRequest" Message="Updating KeyVaultKeyUri is not supported\r\nActivityId: 5f367e7d-92ee-42a7-966b-fb12e53ab69a, Microsoft.Azure.Documents.Common/2.11.0"`

See https://github.com/Azure/azure-rest-api-specs/issues/10323

But I will add a test for it overall

File: azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go [hunk 1]
```diff
@@ -993,3 +1016,60 @@ func checkAccAzureRMCosmosDBAccount_basic(data acceptance.TestData, consistency
 		resource.TestCheckResourceAttrSet(data.ResourceName, "secondary_readonly_key"),
 	)
 }
+
```
Comment (petems) [inline azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go:1019]: @katbyte @jackofallops I've created a test using a key vault example from a different test, can I get some 👀 on it? And run it as an acc-test if it looks correct?

**Comment by petems** [comment]
Just did a manual test, and this config works for me now:

```hcl
provider "azurerm" {
  features {}
}
resource "azurerm_resource_group" "test" {
  name     = "acctestRG-cosmos-123456"
  location = "West US 2"
}

data "azurerm_client_config" "current" {}

data "azuread_service_principal" "cosmosdb" {
  display_name = "Azure Cosmos DB"
}

resource "azurerm_key_vault" "test" {
  name                = "acctestkeyvault123456789"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  tenant_id           = data.azurerm_client_config.current.tenant_id
  sku_name            = "premium"

  purge_protection_enabled = true
  soft_delete_enabled      = true

  access_policy {
    tenant_id = data.azurerm_client_config.current.tenant_id
    object_id = data.azurerm_client_config.current.object_id

    key_permissions = [
      "list",
      "create",
      "delete",
      "get",
      "update",
    ]

    secret_permissions = [
      "get",
      "delete",
      "set",
    ]
  }

  access_policy {
    tenant_id = data.azurerm_client_config.current.tenant_id
    object_id = data.azuread_service_principal.cosmosdb.id

    key_permissions = [
      "list",
      "create",
      "delete",
      "get",
      "update",
      "unwrapKey",
      "wrapKey",
    ]

    secret_permissions = [
      "get",
      "delete",
      "set",
    ]
  }
}

resource "azurerm_key_vault_key" "test" {
  name         = "examplekey123456"
  key_vault_id = azurerm_key_vault.test.id
  key_type     = "RSA"
  key_size     = 2048

  key_opts = [
    "decrypt",
    "encrypt",
    "sign",
    "unwrapKey",
    "verify",
    "wrapKey",
  ]
}

resource "azurerm_cosmosdb_account" "test" {
  name                = "acctest-ca-123456"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  offer_type          = "Standard"
  kind                = "MongoDB"
  key_vault_key_uri   = "${azurerm_key_vault.test.vault_uri}keys/${azurerm_key_vault_key.test.name}/"
  consistency_policy {
    consistency_level = "Strong"
  }
  geo_location {
    location          = azurerm_resource_group.test.location
    failover_priority = 0
  }
}
```

Will fix the test to use this

File: azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go [hunk 2]
```diff
@@ -993,3 +1016,106 @@ func checkAccAzureRMCosmosDBAccount_basic(data acceptance.TestData, consistency
...
+}
+
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_key_vault" "test" {
+  name                = "acctestkeyvault%d"
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+  tenant_id           = data.azurerm_client_config.current.tenant_id
+  sku_name            = "standard"
```
Comment (ArcturusZhang) [inline azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go:1042]: To get the CMK work, we will have to enable `soft_delete` and `purge_protection` (see the document [here](https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-setup-cmk#configure-your-azure-key-vault-instance)). And these two features may requires a premium SKU (not very sure, you could have a try)

Comment (petems) [reply azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go]: Good catch, I did this in my manual test code already and it worked, so I'll update the test here for it

File: website/docs/d/cosmosdb_account.html.markdown [hunk 3]
```diff
@@ -45,6 +45,8 @@ The following attributes are exported:
 
 * `kind` - The Kind of the CosmosDB account.
 
+* `key_vault_key_uri` - The Key Vault URI for encryption
```
Comment (katbyte) [inline website/docs/d/cosmosdb_account.html.markdown:48]:
```suggestion
* `key_vault_key_uri` - The Key Vault key URI for CMK encryption.
```

File: website/docs/r/cosmosdb_account.html.markdown [hunk 4]
```diff
@@ -92,6 +92,8 @@ The following arguments are supported:
 
 * `is_virtual_network_filter_enabled` - (Optional) Enables virtual network filtering for this Cosmos DB account.
 
+* `key_vault_key_uri` - (Optional) A Key Vault URI for encryption eg. `https://<vault-name>.vault.azure.net/keys/<key-name>`
```
Comment (katbyte) [inline website/docs/r/cosmosdb_account.html.markdown:95]:
```suggestion
* `key_vault_key_uri` - (Optional) A Key Vault key URI for CMK encryption in the format of `https://<vault-name>.vault.azure.net/keys/<key-name>`.
```

File: azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go [hunk 5]
```diff
@@ -993,3 +1016,109 @@ func checkAccAzureRMCosmosDBAccount_basic(data acceptance.TestData, consistency
...
+  consistency_policy {
+    consistency_level = "%s"
+  }
+
+  geo_location {
+    location          = azurerm_resource_group.test.location
+    failover_priority = 0
+  }
+
+
```
Comment (katbyte) [inline azurerm/internal/services/cosmos/cosmosdb_account_resource_test.go:1121]:
```suggestion
```

**Review by katbyte — APPROVED** [review]
Thanks @petems - aside from a couple comments this LGTM

**Comment by petems** [comment]
@katbyte All good to merge? I dont have button rights 😄

**Review by katbyte — APPROVED** [review]
LGTM 👍

**Comment by jackofallops** [comment]
Tests passing (1 fail transient, unrelated)
image

**Pull request merged**